File: Code/线性方程组的数值求解.cpp
/*
线性方程组的数值求解一共有两个文件
该文件使用的是逐个元素计算的方法，没有使用矩阵运算
*/
#include <iostream>
#include <stdio.h>
#include <vector>
#include <cmath>
using namespace std;

// 你需要在这里修改方程组的形状和误差限
#define N 4
#define EPS 1e-6

// 在这里修改方程组，注意要与你之前设置的方程组形状对应
// 注意，如果你使用这个文件的话，必须保证 a 主对角线上的元素不为0
const vector<vector<double>> a = {
    {1, 1, 0, 0},
    {0, 1, 0, -2},
    {0, -1, 1, 1},
    {0, -1, 0, 3}
};
const vector<double> b = {1, 1, 1, 1};
// 在这设置初始点
const vector<double> init_point = {1, 1, 1, 1};

vector<double> Jacobi()
{
    cout << "Jacobi" << endl;
    vector<vector<double>> x(2, init_point);
    double e = 1.;
    int cnt = 0;
    while(e > EPS)
    {
        int k = cnt % 2, kk = (cnt + 1) % 2;
        cnt ++;
        for(int i = 0; i < N; i ++)
        {
            x[kk][i] = b[i];
            for(int j = 0; j < N; j ++)
            {
                if(j == i) continue;
                x[kk][i] -= a[i][j] * x[k][j];
            }
            x[kk][i] /= a[i][i];
        }
        e = 0;
        for(int i = 0; i < N; i ++) e += fabs(x[kk][i] - x[k][i]);
        cout << "cnt=" << cnt << " ";
        for(int i = 0; i < N; i ++) cout << x[kk][i] << " ";
        cout << "e=" << e << endl;
    }
    return x[cnt % 2];
}

vector<double> Guass_Seidel()
{
    cout << "Guass_Seidel" << endl;
    vector<vector<double>> x(2, init_point);
    double e = 1.;
    int cnt = 0;
    while(e > EPS)
    {
        int k = cnt % 2, kk = (cnt + 1) % 2;
        cnt ++;
        for(int i = 0; i < N; i ++)
        {
            x[kk][i] = b[i];
            for(int j = 0; j < i; j ++) x[kk][i] -= a[i][j] * x[kk][j];
            for(int j = i + 1; j < N; j ++) x[kk][i] -= a[i][j] * x[k][j];
            x[kk][i] /= a[i][i];
        }
        e = 0;
        for(int i = 0; i < N; i ++) e += fabs(x[kk][i] - x[k][i]);
        cout << "cnt=" << cnt << " ";
        for(int i = 0; i < N; i ++) cout << x[kk][i] << " ";
        cout << "e=" << e << endl;
    }
    return x[cnt % 2];
}

vector<double> Relaxation(double w)
{
    cout << "Relaxation" << endl;
    vector<vector<double>> x(2, init_point);
    double e = 1.;
    int cnt = 0;
    while(e > EPS)
    {
        int k = cnt % 2, kk = (cnt + 1) % 2;
        cnt ++;
        for(int i = 0; i < N; i ++)
        {
            double d = b[i];
            for(int j = 0; j < i; j ++) d -= a[i][j] * x[kk][j];
            for(int j = i; j < N; j ++) d -= a[i][j] * x[k][j];
            x[kk][i] = x[k][i] + w * d / a[i][i];
        }
        e = 0;
        for(int i = 0; i < N; i ++) e += fabs(x[kk][i] - x[k][i]);
        cout << "cnt=" << cnt << " ";
        for(int i = 0; i < N; i ++) cout << x[kk][i] << " ";
        cout << "e=" << e << endl;
    }
    return x[cnt % 2];
}

int main()
{
    vector<double> ans = Jacobi();
    for(int i = 0; i < N; i ++) cout << ans[i] << " ";
    cout << "\n" << endl;

    ans = Guass_Seidel();
    for(int i = 0; i < N; i ++) cout << ans[i] << " ";
    cout << "\n" << endl;

    ans = Relaxation(0.9);
    for(int i = 0; i < N; i ++) cout << ans[i] << " ";
    cout << "\n" << endl;

    return 0;
}

File: Code/线性方程组的数值求解.py
import numpy as np

# 在这了定义你的 Ax=B 以及初值和误差限
# 注意 A 的主对角线上不要有 0
A = [[1, 1, 0, 0],
     [0, 1, 0, -2],
     [0, -1, 1, 1],
     [0, -1, 0, 3]]
B = [1, 1, 1, 1]
INIT = [1, 1, 1, 1]
EPS = 1e-6


def Jacobi(a, b, x0):
    x = np.array([x0 for _ in range(2)], dtype=np.float64)
    a = np.array(a, dtype=np.float64)
    b = np.array(b, dtype=np.float64)
    d = np.diag(np.diag(a))                    
    print("Jacobi")
    cnt, e = 0, 1.
    while e > EPS and cnt < 100:
        k, kk = cnt % 2, (cnt + 1) % 2
        cnt += 1
        x[kk] = -np.linalg.inv(d) @ (((a - d) @ x[k]) - b)
        e = sum(abs(x[k] - x[kk]))
        print(f"{cnt} x =", *x[kk], f"e = {e}")
    return x[cnt % 2]


def Gauss_Seidel(a, b, x0):
    x = np.array([x0 for _ in range(2)], dtype=np.float64)
    a = np.array(a, dtype=np.float64)
    b = np.array(b, dtype=np.float64)
    dl = np.tril(a)                             
    u  = np.triu(a, 1)                          
    print("Gauss-Seidel")
    cnt, e = 0, 1.
    while e > EPS and cnt < 100:
        k, kk = cnt % 2, (cnt + 1) % 2
        cnt += 1
        x[kk] = np.linalg.solve(dl, b - u @ x[k])
        e = sum(abs(x[k] - x[kk]))
        print(f"{cnt} x =", *x[kk], f"e = {e}")
    return x[cnt % 2]


def Relaxation(a, b, x0, omega=0.9):
    x = np.array([x0 for _ in range(2)], dtype=np.float64)
    a = np.array(a, dtype=np.float64)
    b = np.array(b, dtype=np.float64)
    dl = np.tril(a)                             
    u  = np.triu(a, 1)
    cnt, e = 0, 1.
    while e > EPS and cnt < 100:
        k, kk = cnt % 2, (cnt + 1) % 2
        cnt += 1
        x_gs = np.linalg.solve(dl, b - u @ x[k])      
        x[kk] = x[k] + omega * (x_gs - x[k])          
        e = sum(abs(x[k] - x[kk]))
        print(f"{cnt} x =", *x[kk], f"e = {e}")
    return x[cnt % 2]


if __name__ == "__main__":
    ans = Jacobi(A, B, INIT)
    print("Jacobi result:", *ans, end="\n\n")

    ans = Gauss_Seidel(A, B, INIT)
    print("Gauss-Seidel result:", *ans, end="\n\n")

    ans = Relaxation(A, B, INIT, omega=0.9)
    print("Relaxation result:", *ans)
